### PR TITLE
[WIP] Extend Kubeflow Pipelines DSL with support for Databricks operations

### PR DESCRIPTION
**Pull request by magencio**
1st version of kfp.dsl.databricks, a dsl extension to interact with Azure Databricks using the Azure Databricks operator for Kubernetes.
It supports the following operations for now:
* CreateClusterOp, to create a cluster in Databricks.
* DeleteClusterOp, to delete an existing cluster from Databricks.
* CreateJobOp, to create a Spark Job in Databricks.
* DeleteJobOp, to delete an existing Spark Job from Databricks.
* SubmitRunOp, to submit a one-time Run in Databricks.
* DeleteRunOp, to delete an existing Run.

We can create the Ops by using the complete Databricks spec as a Python dictionary or its individual
parts as named parameters (not supported by Job Ops yet). SubmitRunOp also 
provides 2 helper methods to create the Op by using a json string containing the spec or a file name
containing the spec in json format.

### CODE REVIEW

File: sdk/python/kfp/dsl/databricks/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2019 microsoft.
```
Comment (storey247) [inline sdk/python/kfp/dsl/databricks/__init__.py]: Random question: are we actually able to use the MSFT copyright header when contributing back to OSS? I only ask as it seems a bit strange if this gets merged back into the main repo

File: sdk/python/kfp/dsl/databricks/_job_op.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+        if not spec:
+            raise ValueError("You need to provide a spec.")
+
+        super().__init__(
+            k8s_resource = {
+                "apiVersion": "databricks.microsoft.com/v1alpha1",
+                "kind": "Djob",
+                "metadata": {
+                    "name": str(job_name_prefix) + str(time.time()),
```
Comment (storey247) [inline sdk/python/kfp/dsl/databricks/_job_op.py]: why are we appending the time? I understand that job names have to unique, but we don't follow this pattern in the cluster DSL

File: sdk/python/kfp/dsl/databricks/_run_op.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            elif existing_cluster_id and new_cluster:
+                raise ValueError("You cannot provide both a existing_cluster_id and a new_cluster.")
+            elif not existing_cluster_id:
+                spec["new_cluster"] = new_cluster
+            else:
+                spec["existing_cluster_id"] = existing_cluster_id
+
+            if not spark_jar_task and not spark_python_task and not spark_submit_task and not notebook_task:
+                raise ValueError("You need to provide a spark_jar_task or a spark_python_task or a spark_submit_task or a notebook_task.")
+            elif spark_jar_task and not spark_python_task and not spark_submit_task and not notebook_task:
```
Comment (storey247) [inline sdk/python/kfp/dsl/databricks/_run_op.py]: this logic looks a bit strange, personally I would say that the DSL should probably just map the spark tasks and notebook tasks and then let the operator understand if the combination is valid

I fear if this logic is here then it means we have duplication of business concerns between the operator and the DSL

**Review by storey247 — COMMENTED** [review]
A few comments

File: sdk/python/kfp/dsl/databricks/__init__.py (hunk 1 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/__init__.py]: I'm not sure either. I guess it's ok just to remove them for now

File: sdk/python/kfp/dsl/databricks/_job_op.py (hunk 2 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_job_op.py]: Didn't focus too much on Jobs yet. I just got that code from an example pipeline that was using ResourceOp directly. But you are right, we can remove that for consistency.

File: sdk/python/kfp/dsl/databricks/_run_op.py (hunk 3 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_run_op.py]: These business rules are documented in Databricks API. I agree we may be duplicating business rules. But if we fail the checks in the DSL we wouldn't be creating invalid k8s CRDs and making the operator to work for nothing. I was actually thinking about the possibility of doing these checks when we get the whole spec, too. Although implementing and maintaining the Databricks DSL would be much easier if we just don't care what we get from the user and pass it through to the operator.

Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_run_op.py]: I've been doing further tests, and apparently Databricks API won't return any error if the different parts of the spec don't follow the rules specified in the docs. Databricks will try to execute the run anyway. So I guess it's ok to remove these checks from the DSL if the API itself is not enforcing them.

Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_run_op.py]: I removed the additional checks

File: sdk/python/kfp/dsl/databricks/_cluster_op.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+from .._resource_op import ResourceOp
+
+# Azure Databricks Operator and Golang SDK for Databricks types:
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/_cluster_op.py]: it wasn't clear to me what these comments are used for
I looked through the pre-existing DSL classes and couldn't see anything similar
why are these here?

File: sdk/python/kfp/dsl/databricks/_cluster_op.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        if not cluster_name:
+            raise ValueError("You need to provide a cluster_name.")
+
+        if not spec:
+            raise ValueError("You need to provide a spec.")
+
+        super().__init__(
+            k8s_resource = {
+                "apiVersion": "databricks.microsoft.com/v1alpha1",
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/_cluster_op.py:138]: any idea how often/when this apiVersion will change?

File: sdk/python/kfp/dsl/databricks/_cluster_op.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __init__(self,
+        name: str = None,
+        cluster_name: str = None):
+        """Create a new instance of DeleteClusterOp.
+
+        Args:
+
+            name: the name of the op. It does not have to be unique within a pipeline
+                because the pipeline will generates a unique new name in case of conflicts.
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/_cluster_op.py]: `generates` should be `generate`
I'd probably reword as...

`because the pipeline will generate a new unique name in case of a conflict`

File: sdk/python/kfp/dsl/databricks/_cluster_op.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            }, 
+            action = "delete",
+            name = name)
+
+    @property
+    def resource(self):
+        """`Resource` object that represents the `resource` property in
+        `io.argoproj.workflow.v1alpha1.Template`.
+        """
+        return self._resource
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/_cluster_op.py]: have you run this code through a linter/formatter?
just curious, as many linters/formatters prefer there to be a line feed at the end of the last line of code

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 4 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: I added them there for my reference on the expected specs, the status, behaviour of the API, success conditions, etc.

**Comment by martinpeck** [comment]
General comments:

- I wonder if it might be worth creating a README in the `databricks` folder to explain these additional DSL operators, and any setup/testing notes that might be needed

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 5 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py:138]: No, no idea. I guess it depends how often we add new features to the Databricks operator

**Comment by martinpeck** [comment]
Oh, the other thing that's strange is the dates associated with this PR...

image

The commits were 27 days ago? Is that correct, or is this something to do with the force push you made?

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 5 above)
Comment (martinpeck) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py:138]: So, this API version is specific to the Azure Databricks Operator, right?
So, to use these DSL extensions you not only require Databricks, you also need to have deployed the operator?
I suspect that's something we should make clear in some README docs, so that someone trying to run/test these DSL extensions knows the steps required to get things up and running.

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 6 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: I just copied and pasted that from https://github.com/magencio/pipelines/blob/databricks-wrapper/sdk/python/kfp/dsl/_container_op.py#L694. I'll change it

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 4 above)
Comment (martinpeck) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: I wonder if this would be better outside of the code, and in docs (even if it's just a README)

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 6 above)
Comment (martinpeck) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: Ah yes! :) The typo was in the original.

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 7 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: I just realized linter wasn't working on my VS Code instance. I managed to make it work. I'll review the feedback from the tool

**Comment by magencio** [comment]
Yes, there were several small commits, so I squashed them all into a single one

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 4 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: I removed them.

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 6 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: Change done

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 5 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py:138]: Yes, I will add that info to the README docs

Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py:138]: Added README file

File: sdk/python/kfp/dsl/databricks/README.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+These are the supported Ops up to date:
+
+- CreateClusterOp, to create a cluster in Databricks.
+- DeleteClusterOp, to delete an existing cluster from Databricks.
+- CreateJobOp, to create a Spark Job in Databricks.
+- DeleteJobOp, to delete an existing Spark Job from Databricks.
+- SubmitRunOp, to submit a one-time Run in Databricks.
+- DeleteRunOp, to delete an existing Run.
+
+We can create the Ops by using the complete Databricks spec as a Python dictionary or its individual
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/README.md]: Would suggest you change this to:

> For each of these there are two ways a Kubeflow user can create the Ops
> 1. By passing the complete Databricks spec for the Op within a Python Dictionary
> 2. By using named parameters

File: sdk/python/kfp/dsl/databricks/README.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+parts as named parameters (not supported by Job Ops yet). SubmitRunOp also
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/README.md]: Will this be supported before we push the real PR?

File: sdk/python/kfp/dsl/databricks/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
+# Introduction to Databricks DSL 
+
+Databricks DSL provides a set of [Kubeflow Pipeline](https://www.kubeflow.org/docs/pipelines/) Tasks
+(Ops) which let us manipulate [Azure Databricks](https://azure.microsoft.com/services/databricks/) 
+resources using the [Azure Databricks Operator for Kubernetes](
+https://github.com/microsoft/azure-databricks-operator).
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/README.md]: Perhaps add the sentence...

> "This makes the user experience much nicer, and less error prone, than using the ResourceOp to manipulate these Databricks resources.

File: sdk/python/kfp/dsl/databricks/_cluster_op.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
+from .._resource_op import ResourceOp
+
+class CreateClusterOp(ResourceOp):
+    """Represents an op which will be translated into a Databricks Cluster creation resource
```
Comment (martinpeck) [inline sdk/python/kfp/dsl/databricks/_cluster_op.py]: Should we write `Op` rather than `op` here?

Comment (martinpeck) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: If so, you may want to search/replace for them all

**Comment by martinpeck** [comment]
Overall, this PR looks like it's in really good shape. Thanks @magencio!

**Review by martinpeck — APPROVED** [review]
I've left some minor comments regarding text/comments, but this LGTM

File: sdk/python/kfp/dsl/databricks/README.md (hunk 9 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/README.md]: Yes, that is why I changed the title of this PR to add [WIP]. I'm currently working on supporting named params for Job Ops too.

File: sdk/python/kfp/dsl/databricks/README.md (hunk 8 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/README.md]: Done

File: sdk/python/kfp/dsl/databricks/README.md (hunk 9 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/README.md]: Named params are supported in Job Ops now.

File: sdk/python/kfp/dsl/databricks/README.md (hunk 10 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/README.md]: Done.

File: sdk/python/kfp/dsl/databricks/_cluster_op.py (hunk 11 above)
Comment (magencio) [reply sdk/python/kfp/dsl/databricks/_cluster_op.py]: Done

**Comment by magencio** [comment]
As Kubeflow Pipelines team rejected the idea of us modifying the DSL, we finally merged the changes in branch https://github.com/magencio/pipelines/tree/databricks-wrapper-samples instead of the changes in this branch.
